src/typechecker.erl

```diff
@@ -3129,8 +3129,8 @@ type_check_comprehension_in(Env, ResTy, OrigExpr, Compr, Expr, P, [Pred | Quals]
     %% TODO: As a non-boolean predicates don't give runtime errors, we could
     %% possibly add a configuration parameter to toggle this check.
     {VB1, Cs1} = type_check_expr_in(Env, {type, erl_anno:new(0), 'boolean', []}, Pred),
-    {VB2, Cs2} = type_check_comprehension_in(Env, ResTy, OrigExpr, Compr, Expr, P, Quals),
-    {union_var_binds(VB1, VB2, Env), constraints:combine(Cs1, Cs2)}.
+    {VB2, Cs2} = type_check_comprehension_in(union_var_binds([VB1], Env), ResTy, OrigExpr, Compr, Expr, P, Quals),
+    {union_var_binds([VB2], Env), constraints:combine(Cs1, Cs2)}.
```

Comment:
Shouldn't VB1 be returned too? I'm thinking about a comprehension like `[Id || begin Id = 1, true end]`.

```suggestion
    Env1 = union_var_binds([VB1], Env),
    {VB2, Cs2} = type_check_comprehension_in(Env1, ResTy, OrigExpr, Compr, Expr, P, Quals),
    {union_var_binds([VB2], Env1), constraints:combine(Cs1, Cs2)}.
```

Reply:
It was already folded into `VB2`. The following passed on `f` and, as expected, failed on `g`:

```
-spec f() -> [integer()].
f() ->
    [Id || begin Id = 1, true end].

-spec g() -> [atom()].
g() ->
    [Id || begin Id = 1, true end].
```

However, your suggestion made it more explicit, so let's keep it :+1:

Reply:
> It was already folded into VB2.

Are you sure? I though the returned VB2 only contains new var bindings from the remaining generators and filters (Quals). I may be wrong though.

Reply:
This commit - https://github.com/josefs/Gradualizer/pull/444/commits/dd037e5eafad8d2e6b62d415df3a789f438b8d09 - folds `VB1` into `Env` before passing it down to `type_check_comprehension_in`. When we get `VB2` from it, `VB1` is already included.